streaming/receiver.c

```diff
@@ -51,50 +51,230 @@ static void rrdpush_receiver_thread_cleanup(void *ptr) {
+        time_t gap_end   = st->state->window_end;
+        time_t gap_points = (gap_end - expected_t) / st->update_every;
+        if (gap_points > rpt->max_gap)
+            gap_points = rpt->max_gap;
+        time_t gap_first = gap_end - gap_points * st->update_every;
+        if (gap_first > st->state->window_start) {
+            // TODO - put a test scenario in to capture this, triggers a problem in the dbengine
+            debug(D_REPLICATION, "Gap detected on %s: expected %ld, block %ld-%ld. Exceeds max_gap %u, ignoring...",
+                  st->name, (long)expected_t, (long)st->state->window_start, (long)st->state->window_end, rpt->max_gap);
+            skip_gap(st, expected_t, st->state->window_end - st->update_every);
+        } else {
+            if (gap_first > expected_t)
+                skip_gap(st, expected_t, gap_first - st->update_every);
+            debug(D_REPLICATION, "Gap detected on %s: expected %ld, block %ld-%ld. Requesting %ld-%ld",
+                  st->name, (long)expected_t, (long)st->state->window_start, (long)st->state->window_end,
+                  (long)gap_first, (long)gap_end);
+            send_replication_req(user->host, st->id, gap_first, gap_end);
+            st->state->ignore_block = 1;
+            return PARSER_RC_OK;
+        }
+    }
+    user->st->last_updated.tv_sec = user->st->state->window_start;
+    user->st->state->ignore_block = 0;
+    debug(D_REPLICATION, "Replication on %s @ %ld, block %ld-%ld last_update=%ld", st->name, now,
+                         user->st->state->window_start, user->st->state->window_end, st->last_updated.tv_sec);
+    return PARSER_RC_OK;
+disable:
+    errno = 0;
+    error("Replication failed - Invalid REPBEGIN %s %s %s on host %s. Disabling it.", words[1], words[2], words[3],
+          user->host->hostname);
+    user->enabled = 0;
+    return PARSER_RC_ERROR;
+}
+
+PARSER_RC streaming_rep_dim(char **words, void *user_v, PLUGINSD_ACTION *plugins_action) {
+    PARSER_USER_OBJECT *user = user_v;
+    UNUSED(plugins_action);
+    char *id        = words[1];
+    char *time_txt  = words[2];
+    char *value_txt = words[3];
+
+    if (user->st == NULL) {
+        errno = 0;
+        error("Received RRDDIM for %s out of sequence", id);
+        goto disable;
+    }
+
+    if (user->st->state->ignore_block)
+        return PARSER_RC_OK;
+
+    if (!id || !time_txt || !value_txt)
+        goto disable;
+
+    time_t timestamp = (time_t)str2ul(time_txt);
+    storage_number value = str2ull(value_txt);
+
+    RRDDIM *rd = rrddim_find(user->st, id);
+    //time_t st_last = rrdset_last_entry_t(user->st);  UNUSED?
+    if (rd == NULL) {
+        errno = 0;
+        error("Unknown dimension \"%s\" on %s during replication - ignoring", id, user->st->name);
         return PARSER_RC_OK;
     }
+
+    if (user->st->rrd_memory_mode == RRD_MEMORY_MODE_DBENGINE)
+    {
+        rd->state->collect_ops.store_metric(rd, timestamp * USEC_PER_SEC, value);
```

Comment:
You should detect `SN_EMPTY_SLOT` here. Whenever you detect a gap, you flush the metric and you don't call `store_metric()` anymore. This has 2 advantages:
- You don't write gaps to the database. Imagine a 1 month gap for instance.
  - You still transmit the 1-month gap though, which should also be fixed.
- By flushing at the gap, you align the first non-empty data point to a new page, thus supporting `update_every` changes. You must avoid having data points with variable granularity in the same page at all costs:
  - e.g. [1112221111] will just put the data points with the correct starting and ending timestamps, but all intermediate timestamps will be incorrectly inferred)

Reply:
the frequency change must also be inferred by detecting time jumps.

Reply:
I don't think that we should detect SN_EMPTY_SLOT here. We are trying to replicate the state of the collecting node's db. If there is an empty slot at that point in the time-series (which can occur for several different reasons) then we should create one here. We should stick to the case that the `update_every` has changed, which we can infer:
* Inside `rrdeng_store_metric_next()`
*  `spacing = (descr->end_time - decr->start_time) / (descr->page_length / sizeof(number))`
* `if ((point_in_time - desc->end_time) % spacing != 0)` then there is a change.
On a change we should call `rrdeng_store_metric_flush_current_page(rd)` to force a new page after the change in time-frame alignment. This will collapse any gap immediately before a change in `update_every` - which probably does exist because the only valid scenario is restarting the agent.

Reply:
Commit with this change incoming and then I will rerun the testing set in parallel to check it has broken any of the existing cases.

Reply:
This is done, I'll leave you to resolve the conversation when you are happy with the change in `store_next`